File: qizhi/mobile/MentorAI-iOS/MentorAI/Chat/SessionListViewModel.swift
import Foundation
import os

@MainActor
final class SessionListViewModel: ObservableObject {
    @Published private(set) var sessions: [ChatSession] = []
    @Published private(set) var isLoading: Bool = false
    @Published var error: String?

    private let api: SessionAPI
    private let tokenProvider: () -> String?
    private let log = Logger(subsystem: "com.mentorai.app", category: "SessionList")

    init(api: SessionAPI, tokenProvider: @escaping () -> String?) {
        self.api = api
        self.tokenProvider = tokenProvider
    }

    func refresh() async {
        guard let token = tokenProvider(), !token.isEmpty else {
            error = "未登录"
            return
        }
        isLoading = true
        error = nil
        defer { isLoading = false }
        do {
            let list = try await api.list(token: token)
            sessions = list.sorted { $0.sortKey > $1.sortKey }
            log.info("Loaded \(list.count) sessions")
        } catch let err as AuthError {
            error = err.errorDescription
            log.error("Session list failed: \(err.errorDescription ?? "?", privacy: .public)")
        } catch {
            self.error = error.localizedDescription
            log.error("Session list failed: \(error.localizedDescription, privacy: .public)")
        }
    }

    func upsert(_ session: ChatSession) {
        if let idx = sessions.firstIndex(where: { $0.id == session.id }) {
            sessions[idx] = session
        } else {
            sessions.insert(session, at: 0)
        }
        sessions.sort { $0.sortKey > $1.sortKey }
    }

    func delete(id: String) async {
        guard let token = tokenProvider(), !token.isEmpty else {
            error = "未登录"
            return
        }
        guard let idx = sessions.firstIndex(where: { $0.id == id }) else { return }
        let removed = sessions.remove(at: idx)
        do {
            try await api.delete(id: id, token: token)
            log.info("Deleted session \(id, privacy: .public)")
        } catch let err as AuthError {
            sessions.insert(removed, at: idx)
            sessions.sort { $0.sortKey > $1.sortKey }
            error = err.errorDescription
            log.error("Delete failed: \(err.errorDescription ?? "?", privacy: .public)")
        } catch {
            sessions.insert(removed, at: idx)
            sessions.sort { $0.sortKey > $1.sortKey }
            self.error = error.localizedDescription
            log.error("Delete failed: \(error.localizedDescription, privacy: .public)")
        }
    }
}

File: qizhi/mobile/MentorAI-Android/app/src/main/kotlin/com/mentorai/app/chat/SessionListViewModel.kt
package com.mentorai.app.chat

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mentorai.app.networking.AuthError
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Holds the chat-sessions list. Mirrors iOS `SessionListViewModel` — refresh on demand, delete
 * by id with confirmation, surface errors when the list is empty.
 */
class SessionListViewModel(
    private val sessionApi: SessionAPI,
    private val tokenProvider: () -> String?,
) : ViewModel() {

    private val _sessions = MutableStateFlow<List<ChatSession>>(emptyList())
    val sessions: StateFlow<List<ChatSession>> = _sessions.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    fun refresh() {
        val token = tokenProvider() ?: return
        _isLoading.value = true
        _error.value = null
        viewModelScope.launch {
            try {
                // Mirror iOS: newest first by sortKey (update_time, falling back to create_time).
                _sessions.value = sessionApi.list(token).sortedByDescending { it.sortKey }
            } catch (err: AuthError) {
                _error.value = err.errorDescription
            } catch (t: Throwable) {
                _error.value = t.localizedMessage ?: "加载失败"
            } finally {
                _isLoading.value = false
            }
        }
    }

    /** Mirrors iOS `upsert` — replace an existing session or insert a new one, then re-sort. */
    fun upsert(session: ChatSession) {
        val current = _sessions.value
        val idx = current.indexOfFirst { it.id == session.id }
        val updated = if (idx >= 0) {
            current.toMutableList().also { it[idx] = session }
        } else {
            current.toMutableList().also { it.add(0, session) }
        }
        _sessions.value = updated.sortedByDescending { it.sortKey }
    }

    /**
     * Optimistic delete, mirroring iOS `SessionListViewModel.delete`: drop the row immediately,
     * then call the API. On failure, re-insert it at the original index, re-sort, and surface the
     * error.
     */
    fun delete(id: String) {
        val token = tokenProvider()
        if (token.isNullOrBlank()) {
            _error.value = "未登录"
            return
        }
        val idx = _sessions.value.indexOfFirst { it.id == id }
        if (idx < 0) return
        val removed = _sessions.value[idx]
        _sessions.value = _sessions.value.toMutableList().also { it.removeAt(idx) }
        viewModelScope.launch {
            try {
                sessionApi.delete(id, token)
            } catch (err: AuthError) {
                _sessions.value = _sessions.value.toMutableList()
                    .also { it.add(idx, removed) }
                    .sortedByDescending { it.sortKey }
                _error.value = err.errorDescription
            } catch (t: Throwable) {
                _sessions.value = _sessions.value.toMutableList()
                    .also { it.add(idx, removed) }
                    .sortedByDescending { it.sortKey }
                _error.value = t.localizedMessage ?: "删除失败"
            }
        }
    }
}
